crates/kornia-pnp/src/refine.rs

```diff
@@ -0,0 +1,333 @@
+    let mut residuals_p = vec![0.0f32; RESIDUAL_DIM * n];
+    let mut residuals_m = vec![0.0f32; RESIDUAL_DIM * n];
+
+    let project_all_in_place = |x: &[f32; 6], out: &mut [f32]| -> f32 {
+        let r_mat = SO3::exp(Vec3A::from_array([x[0], x[1], x[2]])).matrix();
+        let t_vec = Vec3::new(x[3], x[4], x[5]);
+
+        let mut sum_sq = 0.0f32;
+        for (i, (pw_arr, &uv)) in points_world.iter().zip(points_image.iter()).enumerate() {
+            let pw = Vec3::from_array(*pw_arr);
+            let pc = r_mat * pw + t_vec;
+            let inv_z = 1.0 / pc.z;
+            let u_hat = intr_x.dot(pc) * inv_z;
+            let v_hat = intr_y.dot(pc) * inv_z;
+            let du = u_hat - uv[0];
+            let dv = v_hat - uv[1];
+            out[2 * i] = du;
+            out[2 * i + 1] = dv;
+            sum_sq += du.mul_add(du, dv * dv);
+        }
+        sum_sq
+    };
+
+    let mut lambda = params.lambda_init;
+    let mut err_sq_base = project_all_in_place(&x, &mut residuals);
+
+    let mut iters = 0usize;
+    let mut converged = false;
+
+    // Preallocate J, A and b once
+    let mut j = vec![0.0f32; RESIDUAL_DIM * n * SOLVER_DIM];
+    let mut a = [0.0f32; JTJ_SIZE];
+    let mut b = [0.0f32; SOLVER_DIM];
+
+    while iters < params.max_iters {
+        iters += 1;
+        // Reset accumulators
+        j.fill(0.0);
+        a.fill(0.0);
+        b.fill(0.0);
+        let t_scale = x[3]
+            .abs()
+            .max(x[4].abs())
+            .max(x[5].abs())
+            .max(MIN_TRANSLATION_SCALE);
+        let h_trans = params.fd_trans_base_step * t_scale; // world units
+
+        for k_idx in 0..SOLVER_DIM {
+            // Central differences
+            let h = if k_idx < 3 {
+                params.fd_rot_step
+            } else {
+                h_trans
+            };
+            let mut x_plus = x;
+            let mut x_minus = x;
+            x_plus[k_idx] += h;
+            x_minus[k_idx] -= h;
+            let _ = project_all_in_place(&x_plus, &mut residuals_p);
+            let _ = project_all_in_place(&x_minus, &mut residuals_m);
+            for i in 0..(RESIDUAL_DIM * n) {
+                j[i * SOLVER_DIM + k_idx] = (residuals_p[i] - residuals_m[i]) / (2.0 * h);
+            }
+        }
+
+        // Build normal equations: (J^T J + lambda I) delta = -J^T r
+        for r_i in 0..(RESIDUAL_DIM * n) {
+            let r_val = residuals[r_i];
+            for c in 0..SOLVER_DIM {
+                let j_ic = j[r_i * SOLVER_DIM + c];
+                b[c] += j_ic * r_val;
+                for d in 0..SOLVER_DIM {
+                    a[c * SOLVER_DIM + d] += j_ic * j[r_i * SOLVER_DIM + d];
+                }
+            }
+        }
+        // Damping
+        for d in 0..SOLVER_DIM {
+            a[d * SOLVER_DIM + d] += lambda;
+        }
+
+        // Solve A delta = -b
+        let mut rhs = [0.0f32; SOLVER_DIM];
+        for i in 0..SOLVER_DIM {
```

Comment:
why not use iterators with zip, it's more rust idiomatic and prevents random access every time